gnomad_hail/utils/relatedness.py

```diff
@@ -0,0 +1,402 @@
+        """
+        fam_id = 1  # stores the current family id
+        s_fam = dict()  # stores the family id for each sample
+        fams = defaultdict(list)
+        for pair in parent_child_pairs:
+            if pair[0] in s_fam:
+                if pair[1] in s_fam:
+                    if s_fam[pair[0]] != s_fam[pair[1]]:  # If both samples are in different families, merge the families
+                        new_fam_id = s_fam[pair[0]]
+                        fam_id_to_merge = s_fam[pair[1]]
+                        for s in s_fam:
+                            if s_fam[s] == fam_id_to_merge:
+                                s_fam[s] = new_fam_id
+                        fams[new_fam_id].extend(fams.pop(fam_id_to_merge))
+                else:
+                    s_fam[pair[1]] = s_fam[pair[0]]
+                fams[s_fam[pair[0]]].append(pair)
+            elif pair[1] in s_fam:
+                s_fam[pair[0]] = s_fam[pair[1]]
+                fams[s_fam[pair[1]]].append(pair)
+            else:
+                s_fam[pair[0]] = fam_id
+                s_fam[pair[1]] = fam_id
+                fams[fam_id].append(pair)
+                fam_id += 1
+
+        return list(fams.values())
+
+    def get_trios(
+            fam_id: str,
+            parent_child_pairs: List[Tuple[str, str]],
+            related_pairs: Dict[Tuple[str, str], str]
+    ) -> List[hl.Trio]:
+        """
+        Generates trios based from the list of parent-child pairs in the family and
+        all related pairs in the data. Only complete parent/offspring trios are included in the results.
+
+        The trios are assembled as follows:
+        1. All pairs of unrelated samples with different sexes within the family are extracted as possible parent pairs
+        2. For each possible parent pair, a list of all children is constructed (each child in the list has a parent-offspring pair with each parent)
+        3. If there are multiple children for a given parent pair, all children should be siblings with each other
+        4. Check that each child was only assigned a single pair of parents. If a child is found to have multiple parent pairs, they are ALL discarded.
+
+        :param fam_id: The family ID
+        :param parent_child_pairs: The parent-child pairs for this family
+        :param related_pairs: All related sample pairs in the data
+        :return: List of trios in the family
+        """
+
+        def get_possible_parents(samples: List[str]) -> List[Tuple[str, str]]:
+            """
+            1. All pairs of unrelated samples with different sexes within the family are extracted as possible parent pairs
+
+            :param samples: All samples in the family
+            :return: Possible parent pairs
+            """
+            possible_parents = []
+            for i in range(len(samples)):
+                for j in range(i, len(samples)):
```

Comment:
"for j in range(i + 1, len(samples)): " added +1 because otherwise compares each sample to itself